### ISSUE TITLE
how to understand prediction in SequenceClassificationBRNNTest (classification) 

### ISSUE THREAD

**Issue by kathy-lee** [mlpack/mlpack#2160]
Hi,
in SequenceClassificationBRNNTest (line78) of recurrent_network_test.cpp, I got problem with understanding the prediction. The input is (1,12,10),from "model.Predict(input, prediction)" got prediction is (10,12,10). Print out  prediction.slice(rho-1).col(i=0) is like this: 
  0
  -13.1190
  -15.2380
  -15.7081
  -16.2024
  -15.9800
  -15.5745
  -15.7167
  -16.1695
  -15.8157
I can understand slice(rho-1) means the latest result, but what do these 10 values mean? And where is the final binary classification result? Could anyone help me to make this clear? Many Thanks!

**Comment by zoq** [comment]
Have to refresh my memory, so I'm going to respond tomorrow.

**Comment by zoq** [comment]
Just checked the code, and what you see here is the merged output of the last layer (`model.Add<Linear<> >(4, 5)`) in both directions, this is why the output is twice the size of the last layer. However there are two things that I think aren't necessarily correct; the output size of the last layer should match the number of classes, in this case, 2, there is no problem to use more as done here, but the information isn't useful. Another issue with the test is that we only check the forward pass, so this is basically as I would use a rnn, since the last half of the output vector does contain the information form the backward path.

Hopefully, anything I said above makes sense. I'll see if I can address the issues I mentioned above in the next days.

**Comment by kathy-lee** [comment]
many thanks for your reply, @zoq . Not sure if I get it correctly, do you mean that the output size of the last layer can be different from the number of classes? I could not really understand this, because generally for classification problem, people would set the size of last layer as the number of classes + 1. Shouldn't there be one more layer after  `model.Add<Linear<> >(4, 5)` ? That's also why I am confused with getting `predictionValue` by finding max of `prediction.slice(rho-1).col(i)`. 
My goal is to implement a LSTM network for classification using MLPACK  in c++ environment, while in Keras it's something like this:

```
model = Sequential()
model.add(Bidirectional(LSTM(64, return_sequences=False, stateful=False,input_shape=(60, 640) )))
model.add(Dropout(.5,name='dropout_1'))
model.add(Dense(128, activation='relu', name='DENSE_1'))
model.add(Dropout(.5,name='dropout_2'))
model.add(Dense(5, activation='softmax', name = 'output')) # the number of classes is 5.
```

With MLPACK could I create a `BRNN<>` model object and add corresponding layers as above?

**Comment by kathy-lee** [comment]
By the way, at line https://github.com/mlpack/mlpack/blob/dcfcff4a91bf93f1486a5716b222263153fea1b4/src/mlpack/tests/recurrent_network_test.cpp#L131, do you think the varname should be "accuracy" instead of "error" according to the function context?

**Comment by zoq** [comment]
>  Not sure if I get it correctly, do you mean that the output size of the last layer can be different from the number of classes? I could not really understand this, because generally for classification problem, people would set the size of last layer as the number of classes + 1.

Yes, there is no reason to add more units to the last layer as the number of classes, I was just pointing out that in this case we have more units in the last layer as classes, and this is why the output size is 10.

> That's also why I am confused with getting `predictionValue` by finding max of `prediction.slice(rho-1).col(i)`

Right, the network returns a probability vector and not the final classification (scalar) which is the index of the max value.

> , do you think the varname should be "accuracy" instead of "error" according to the function context?

Yes, good point, using `accuracy` does make more sense, we should change that.

> With MLPACK could I create a BRNN<> model object and add corresponding layers as above?

That should be possible, do you have a simple/minimal dataset for the example above as well? If I patch the test I can write another one with the model you mentioned above, I guess we don't have a test that uses the LSTM layer in combination with the `BRNN` class.

**Comment by kathy-lee** [comment]
That's much clearer, thanks @zoq . It would be great if you can patch a test that uses the LSTM layer with BRNN class. I prepared a tiny dataset serialized from arma::cube to hdf5 format, please let me know if they are okay or not. The training data size is (25,60,640) corresponding to (number of samples, time steps, ESF feature), training label is a arma::vec with a length of 25. 
[test_data001.zip](https://github.com/mlpack/mlpack/files/4100650/test_data001.zip)
[test_labels001.zip](https://github.com/mlpack/mlpack/files/4100651/test_labels001.zip)
[training_data001.zip](https://github.com/mlpack/mlpack/files/4100652/training_data001.zip)
[training_labels001.zip](https://github.com/mlpack/mlpack/files/4100653/training_labels001.zip)

**Comment by zoq** [comment]
Thanks, will write another test based on the dataset.

**Comment by kathy-lee** [comment]
Hi @zoq, I wrote some code according to other tests in recurrent_network_test.cpp, but not sure if it has problems:
`  const int n_timesteps = 60;
    const int n_samples = 25;
    const int n_feature = 640;
    arma::cube prediction;
    arma::cube labels = arma::zeros<arma::cube>(1, n_samples, n_timesteps);
    for(size_t i = 0; i < n_samples; i++) {
          const int value = training_labels[i] + 1;
          labels.tube(0,i).fill(value);
    }

    BRNN<> model(n_samples);
    model.Add<IdentityLayer<>>();
    model.Add<LSTM<>>(n_feature, 64, n_samples);
    model.Add<Dropout<>>(0.1);
    model.Add<Linear<>>(64, 5);
    ens::SGD<> sgd(0.01, 1, 100 );
    model.Train(training_data, labels, sgd);
    model.Predict(test_data, prediction);`
1. I set the dimension of "labels" to arma::cube (1, n_samples, n_timesteps) which is got from the original arma::vec "training_Labels".
2. There is no error with compiling or linking, but it has been running for 10 hours for only n_samples = 25 and not finished yet.
I'm not sure if either the testbed is too big, or something is missing or wrong in my above steps.

**Comment by saksham189** [comment]
> Just checked the code, and what you see here is the merged output of the last layer (model.Add<Linear<> >(4, 5)) in both directions, this is why the output is twice the size of the last layer. However there are two things that I think aren't necessarily correct; the output size of the last layer should match the number of classes, in this case, 2, there is no problem to use more as done here, but the information isn't useful. Another issue with the test is that we only check the forward pass, so this is basically as I would use a rnn, since the last half of the output vector does contain the information form the backward path.

[The RNN test for the noisy sine waves](https://github.com/mlpack/mlpack/blob/master/src/mlpack/tests/recurrent_network_test.cpp#L150) below that also has an output size of 10. Does that also need to be changed?

**Comment by saksham189** [comment]
>  It would be great if you can patch a test that uses the LSTM layer with BRNN class. I prepared a tiny dataset serialized from arma::cube to hdf5 format, please let me know if they are okay or not. The training data size is (25,60,640) corresponding to (number of samples, time steps, ESF feature), training label is a arma::vec with a length of 25.

Hi, @kathy-lee you look here for an example of using brnn with the lstm class. 
https://github.com/mlpack/mlpack/blob/master/src/mlpack/tests/recurrent_network_test.cpp#L693
Let me know if you need any further help :)
